### PR TITLE
Optimize validation tests

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
- Add type declarations
- Remove unnecessary multilines, whitespaces, comments
- Use generators in data providers (for efficient memory consumption)
- Move database related rules testing into own file (to reduce DB load on non-DB tests)
- Group several related tests using a data provider (for scalability of testing)

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: tests/system/Validation/CreditCardRulesTest.php [hunk 1]
```diff
@@ -1220,32 +1205,26 @@ public function creditCardProvider()
...
      */
-    protected function generateCardNum(int $prefix, int $length)
+    private function generateCardNum(int $prefix, int $length): string
     {
         $pos        = mb_strlen($prefix);
         $finalDigit = 0;
         $sum        = 0;
 
-        // Fill in the first values of the string based on $prefix
         $string = str_split($prefix);
```
Comment (kenjis) [inline tests/system/Validation/CreditCardRulesTest.php:1215]: The `$string` is an array. `$digits` is better?

File: tests/system/Validation/CreditCardRulesTest.php [hunk 2]
```diff
@@ -1220,32 +1205,26 @@ public function creditCardProvider()
...
-
-        // Pad out the array to the appropriate length
         $string = array_pad($string, $length, 0);
 
-        // Fill all of the remaining values with random numbers, except the last one.
         while ($pos < $length - 1) {
             $string[$pos++] = random_int(0, 9);
         }
 
-        // Calculate the Luhn checksum of the current values.
```
Comment (kenjis) [inline tests/system/Validation/CreditCardRulesTest.php:1243]: This comment helps code readers. I hope not to remove.

Comment (kenjis) [reply tests/system/Validation/CreditCardRulesTest.php:1243]: Or is it better that extracting method to calculate the Luhn checksum?
Then the code would be like:
```php
$string[--$length] = $this->calculateLuhnChecksum(...);
```

**Comment by MGatner** [comment]
All good rubrics and the tests are passing so I won't go through it all. I do like @kenjis suggestions. Mention me if you need anything else.

File: tests/system/Validation/CreditCardRulesTest.php (hunk 1 above)
Comment (paulbalandan) [reply tests/system/Validation/CreditCardRulesTest.php]: Done

File: tests/system/Validation/CreditCardRulesTest.php (hunk 2 above)
Comment (paulbalandan) [reply tests/system/Validation/CreditCardRulesTest.php:1243]: Extracted

**Review by MGatner — APPROVED** [review]

**Review by kenjis — APPROVED** [review]

**Pull request merged**